hpx/parallel/util/loop.hpp

```diff
@@ -68,12 +69,176 @@ namespace hpx { namespace parallel { namespace util
+            {
+                return lhs - rhs;
+            }
+
+
+            inline bool operator==(const prefetching_iterator& rhs) const {return idx == rhs.idx;}
+            inline bool operator!=(const prefetching_iterator& rhs) const {return idx != rhs.idx;}
+            inline bool operator>(const prefetching_iterator& rhs) const {return idx > rhs.idx;}
+            inline bool operator<(const prefetching_iterator& rhs) const {return idx < rhs.idx;}
+            inline bool operator>=(const prefetching_iterator& rhs) const {return idx >= rhs.idx;}
+            inline bool operator<=(const prefetching_iterator& rhs) const {return idx <= rhs.idx;}
+
+
+            inline std::size_t operator*() const {return idx;}
+        };
+
+
+        //Helper class to initialize prefetching_iterator
+        template<typename T>
+        struct prefetcher_context
+        {
+            std::vector<std::size_t> range;
+            std::vector<std::size_t>::iterator it_begin;
+            std::vector<std::size_t>::iterator it_end;
+            std::size_t chunk_size;
+            std::vector< T * > m;
+            std::size_t range_size;
+
+
+            explicit prefetcher_context (std::size_t begin, std::size_t end,
+                std::size_t p_factor, std::initializer_list< T * > &&l)
+            {
+                std::size_t vector_size = end - begin + 1;
+                range.resize(vector_size);
+                for(std::size_t i=0; i<vector_size; ++i) range[i] = i + begin;
+                it_begin = range.begin();
+                it_end = range.begin() + vector_size - 1;
+                chunk_size = p_factor * 64ul / sizeof(T);
+                m = l;
+                range_size = end + 1;
+            }
+
+            explicit prefetcher_context (std::size_t begin, std::size_t end,
+                std::initializer_list< T * > &&l)
+            {
+                std::size_t vector_size = end - begin + 1;
+                range.resize(vector_size);
+                for(std::size_t i=0; i<vector_size; ++i) range[i] = i + begin;
+                it_begin = range.begin();
+                it_end = range.begin() + vector_size - 1;
+                chunk_size = 64ul / sizeof(T);
+                m = l;
+                range_size = end + 1;
+            }
+
+            prefetching_iterator<T> begin()
+            {
+                return prefetching_iterator<T>(0ul, it_begin, chunk_size, range_size, m);
+            }
+
+            prefetching_iterator<T> end()
+            {
+                return prefetching_iterator<T>(range_size, it_end, chunk_size, range_size, m);
+            }
+
+        };
+
+
+        //function which initialize prefetcher_context
+        template<typename T>
+        prefetcher_context<T> make_prefetcher_context(std::size_t idx_begin, std::size_t idx_end,
+            std::initializer_list< T * > &&l, std::size_t p_factor = 0)
```

Comment:
I don't like this API of having a pointer + size, couldn't we use something like array_view instead?